consensus/raft/raft.go

```diff
@@ -427,9 +435,98 @@ func (rw *raftWrapper) Peers() ([]string, error) {
 	return ids, nil
 }
 
-// only call when Raft is shutdown
-func (rw *raftWrapper) Clean() error {
-	dbh := newDataBackupHelper(rw.dataFolder)
+// latestSnapshot looks for the most recent raft snapshot stored at the
+// provided basedir.  It returns a boolean indicating if any snapshot is
+// readable, the snapshot's metadata, and a reader to the snapshot's bytes
+func latestSnapshot(raftDataFolder string) (bool, *hraft.SnapshotMeta, io.ReadCloser, error) {
+	store, err := hraft.NewFileSnapshotStore(raftDataFolder, RaftMaxSnapshots, nil)
+	if err != nil {
+		return false, nil, nil, err
+	}
+	snapMetas, err := store.List()
+	if err != nil {
+		return false, nil, nil, err
+	}
+	if len(snapMetas) == 0 { // no error if snapshot isn't found
+		return false, nil, nil, nil
+	}
+	meta, r, err := store.Open(snapMetas[0].ID)
+	if err != nil {
+		return false, nil, nil, err
+	}
+	return true, meta, r, nil
+}
+
+// LastStateRaw returns the bytes of the last snapshot stored, its metadata,
+// and a flag indicating whether any snapshot was found.
+func LastStateRaw(cfg *Config) ([]byte, *hraft.SnapshotMeta, bool, error) {
+	// Read most recent snapshot
+	dataFolder, err := makeDataFolder(cfg.BaseDir, cfg.DataFolder)
+	if err != nil {
+		return nil, nil, false, err
+	}
+	found, meta, r, err := latestSnapshot(dataFolder)
+	if err != nil {
+		return nil, nil, false, err
+	}
+	if !found {
+		return nil, nil, false, nil
+	}
+
+	rawBytes, err := ioutil.ReadAll(r)
+	if err != nil {
+		return nil, meta, true, err
+	}
+	return rawBytes, meta, true, nil
+}
+
+// SnapshotSave saves the provided state to a snapshot in the
+// raft data path.  Old raft data is backed up and replaced
+// by the new snapshot
+func SnapshotSave(cfg *Config, newState state.State, meta *hraft.SnapshotMeta) error {
```

Comment:
It certainly could.  I was addressing your earlier comment that `SnapshotSave` unnecessarily duplicates a lot of `LastStateRaw`'s code.  If we don't pass in meta, `SnapshotSave` will duplicate most of `LastStateRaw`.

We could choose not to receive the meta return value from `LastStateRaw` in main, change `SnapshotSave` to not take in meta, and then call `LastStateRaw` this time receiving meta from within `SnapshotSave`.  It would still keep the meta return argument but meta wouldn't be exposed in main.  

I'm pretty neutral about any position in this tradeoff space so let me know what you prefer.

Reply:
maybe make a private LastStateRaw-like function which provides the meta, and call from both LastStateRaw and SnapshotSave ?